### PR TITLE
Address Flaky Tests

### PR DESCRIPTION
**Pull request by anshradh**
## Description
Fix issues with some flaky tests, addressing #551. Most of these were flaky when run in parallel so their flakiness wasn't showing up in CI
## Testing
Ran new tests locally with `pytest-flakefinder` for many (e.g. 500) runs

### CODE REVIEW

**Comment by anshradh** [comment]
@AdamGleave I think this PR is reasonable to merge in now, but also happy to wait in case I find some other candidate tests we want to include - these are the ones that have a > 2% flake rate when run in parallel.

File: tests/algorithms/test_preference_comparisons.py [hunk 1]
```diff
@@ -71,6 +71,21 @@ def agent_trainer(agent, reward_net, venv, rng):
     return preference_comparisons.AgentTrainer(agent, reward_net, venv, rng)
 
 
+def assert_info_arrs_equal(arr1, arr2):  # pragma: no cover
```
[the remark this answers is not in the document]
Comment (anshradh) [reply tests/algorithms/test_preference_comparisons.py:74]: Case 1 is right here, we should be asserting that items1 and 2 are dictionaries!

File: tests/policies/test_exploration_wrapper.py [hunk 2]
```diff
@@ -90,37 +90,65 @@ def test_switch_prob(rng):
...
-    assert num_constant == 1000
+
+    num_randoms_prob_one = []
+    num_constants_prob_one = []
+    num_randoms_prob_half = []
+    num_constants_prob_half = []
+    num_randoms_prob_zero = []
+    num_constants_prob_zero = []
+
+    for s in range(5):
```
[the remark this answers is not in the document]
Comment (anshradh) [reply tests/policies/test_exploration_wrapper.py:101]: Yeah the specific reason I went for this route was because Windows has a different RNG and the original seed we were using caused the test to fail only on Windows, so this was a way to check that that seed was the actual issue. I think probably the better thing to do here is to assert that the mean is > 450 across the five runs with different seeds.

File: tests/policies/test_exploration_wrapper.py [hunk 3]
```diff
@@ -90,37 +90,65 @@ def test_switch_prob(rng):
...
+        np.random.seed(s)
+        obs = np.random.rand(100, 2)
+        for action in wrapper(obs):
+            assert venv.action_space.contains(action)
+            assert wrapper.current_policy == policy
```
[the remark this answers is not in the document]
Comment (anshradh) [reply tests/policies/test_exploration_wrapper.py]: Yeah moved this out, thanks!

File: tests/policies/test_exploration_wrapper.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (dan-pandori) [reply tests/policies/test_exploration_wrapper.py]: Isn't asserting that the mean is between 450 and 550 mathematically equivalent to having one run that's 5 times as long, and asserting the total is between 2250 and 2750? Admittedly, that's ignoring the unlucky seed issue. 

I also thought that there was a push to try to remove static seed usage in these tests in general.

[the remark this answers is not in the document]
Comment (anshradh) [reply tests/policies/test_exploration_wrapper.py]: I agree with you, removed dependence on seeds entirely and increased the number of steps.

File: experiments/bc_benchmark.sh [hunk 4]
```diff
@@ -11,7 +11,8 @@ source experiments/common.sh
 
 ENVS=(seals_cartpole)
 SEEDS=(0 1 2)
-OUTPUT_DIR="output/bc_benchmark/${TIMESTAMP}"
+# To prevent race conditions, we use a different output dir for each process id.
+OUTPUT_DIR="output/bc_benchmark/${TIMESTAMP}/$BASHPID"
```
Comment (AdamGleave) [inline experiments/bc_benchmark.sh:15]: Preference for this being the same directory so we don't create lots of directories with a single subdirectory.

Also suggest using `${}` for consistency with surrounding code although in this situation it doesn't matter.

```suggestion
OUTPUT_DIR="output/bc_benchmark/${TIMESTAMP}-${BASHPID}"
```

File: experiments/dagger_benchmark.sh [hunk 5]
```diff
@@ -4,7 +4,8 @@ source experiments/common.sh
 
 ENVS=(seals_cartpole)
 SEEDS=(0 1 2 3 4)
-LOG_ROOT="output/dagger_benchmark/${TIMESTAMP}"
+# To prevent race conditions, we use a different log root for each process id.
+LOG_ROOT="output/dagger_benchmark/${TIMESTAMP}/$BASHPID"
```
Comment (AdamGleave) [inline experiments/dagger_benchmark.sh:8]: as above
```suggestion
LOG_ROOT="output/dagger_benchmark/${TIMESTAMP}-${BASHPID}"
```

File: experiments/imit_benchmark.sh [hunk 6]
```diff
@@ -11,7 +11,8 @@ source experiments/common.sh
 
 SEEDS=(0 1 2 3 4)
 CONFIG_CSV="experiments/imit_benchmark_config.csv"
-LOG_ROOT="output/imit_benchmark/${TIMESTAMP}"
+# To prevent race conditions, we use a different log root for each process id.
+LOG_ROOT="output/imit_benchmark/${TIMESTAMP}/$BASHPID"
```
Comment (AdamGleave) [inline experiments/imit_benchmark.sh:15]: as above
```suggestion
LOG_ROOT="output/imit_benchmark/${TIMESTAMP}-${BASHPID}"
```

File: experiments/rollouts_from_policies.sh [hunk 7]
```diff
@@ -16,7 +16,8 @@ set -e  # Exit on error.
 source experiments/common.sh
 
 CONFIG_CSV=${CONFIG_CSV:-experiments/rollouts_from_policies_config.csv}
-OUTPUT_DIR="output/train_experts/${TIMESTAMP}"
+# To prevent race conditions, we use a different output dir for each process id.
+OUTPUT_DIR="output/train_experts/${TIMESTAMP}/$BASHPID"
```
Comment (AdamGleave) [inline experiments/rollouts_from_policies.sh:20]:
```suggestion
OUTPUT_DIR="output/train_experts/${TIMESTAMP}-${BASHPID}"
```

File: experiments/transfer_learn_benchmark.sh [hunk 8]
```diff
@@ -8,7 +8,8 @@ source experiments/common.sh
 SEEDS=(0 1 2)
 CONFIG_CSV="experiments/imit_benchmark_config.csv"
 REWARD_MODELS_DIR="data/reward_models"
-LOG_ROOT="output/train_experts/${TIMESTAMP}"
+# To prevent race conditions, we use a different log root for each process id.
+LOG_ROOT="output/train_experts/${TIMESTAMP}/$BASHPID"
```
Comment (AdamGleave) [inline experiments/transfer_learn_benchmark.sh:12]:
```suggestion
LOG_ROOT="output/train_experts/${TIMESTAMP}-${BASHPID}"
```

File: experiments/transfer_learn_benchmark.sh [hunk 9]
```diff
@@ -67,15 +68,18 @@ done
 
 if [[ $NEED_TEST_FILES == "true" ]]; then
   # Generate quick reward models for test.
-  save_dir=tests/testdata/reward_models/${ALGORITHM}
+  # To prevent race conditions, we use a different save_dir for each process id.
+  save_dir=tests/testdata/reward_models/${ALGORITHM}/${TIMESTAMP}/$BASHPID
```
Comment (AdamGleave) [inline experiments/transfer_learn_benchmark.sh:72]:
```suggestion
  save_dir=tests/testdata/reward_models/${ALGORITHM}/${TIMESTAMP}-${BASHPID}
```

File: experiments/transfer_learn_benchmark.sh [hunk 10]
```diff
@@ -67,15 +68,18 @@ done
...
 
   # Wipe directories for writing later.
   if [[ -d ${save_dir} ]]; then
-    rm -r ${save_dir}
+    echo "Wiping ${save_dir}"
```
Comment (AdamGleave) [inline experiments/transfer_learn_benchmark.sh:76]: do we still want this or was it just for debugging?

File: experiments/transfer_learn_benchmark.sh [hunk 11]
```diff
@@ -88,7 +92,7 @@ parallel -j 25% --header : --results "${LOG_ROOT}/parallel/" --colsep , --progre
   '{env_config_name}' seed='{seed}' \
   common.log_dir="${LOG_ROOT}/${ALGORITHM}/{env_config_name}_{seed}/n_expert_demos_{n_expert_demos}" \
   reward_type="RewardNet_unshaped" \
-  reward_path="${REWARD_MODELS_DIR}/${ALGORITHM}/{env_config_name}_0/n_expert_demos_{n_expert_demos}/checkpoints/final/reward_test.pt" \
+  reward_path="${REWARD_MODELS_DIR}/${ALGORITHM}/${TIMESTAMP}/$BASHPID/{env_config_name}_0/n_expert_demos_{n_expert_demos}/checkpoints/final/reward_test.pt" \
```
Comment (AdamGleave) [inline experiments/transfer_learn_benchmark.sh:95]:
```suggestion
  reward_path="${REWARD_MODELS_DIR}/${ALGORITHM}/${TIMESTAMP}-${BASHPID}/{env_config_name}_0/n_expert_demos_{n_expert_demos}/checkpoints/final/reward_test.pt" \
```

File: tests/algorithms/test_preference_comparisons.py [hunk 12]
```diff
@@ -72,6 +72,21 @@ def agent_trainer(agent, reward_net, venv, rng):
...
+    for item1, item2 in zip(arr1, arr2):
+        assert isinstance(item1, dict)
+        assert isinstance(item2, dict)
+        for key, val in item1.items():
+            assert key in item2
+            if isinstance(val, dict):
```
Comment (AdamGleave) [inline tests/algorithms/test_preference_comparisons.py:81]: Do we want to enforce that dicts are only nested one layer deep? If not we could just turn this into a recursive call (it'd also simplify the code)?

**Review by AdamGleave — COMMENTED** [review]
Some fairly nitpicky comments, otherwise it LGTM. Thanks for making our tests reliable Ansh!

File: experiments/transfer_learn_benchmark.sh (hunk 10 above)
Comment (anshradh) [reply experiments/transfer_learn_benchmark.sh:76]: Good catch, thanks!

File: tests/algorithms/test_preference_comparisons.py (hunk 12 above)
Comment (anshradh) [reply tests/algorithms/test_preference_comparisons.py:81]: I think it might make sense to leave as is, so we can enforce that dicts are either unnested or nested one layer deep, so that the test breaking will alert us that the info array contains dicts nested even further? I don't have a strong opinion here though.

Comment (anshradh) [reply tests/algorithms/test_preference_comparisons.py]: Yeah on reflection changed this to the recursive call, makes it easier to reason about.
